File: files/300-399/390/CH-390.R
library(tidyverse)
library(readxl)
library(anytime)
library(lubridate)

path <- "300-399/390/CH-390 Table Transformation.xlsx"
input <- read_excel(path, range = "B3:E6")
test <- read_excel(path, range = "G3:I9")

result <- input %>%
  pivot_longer(-col1, names_to = "col", values_to = "val") %>%
  pivot_wider(names_from = col1, values_from = val) %>%
  separate_longer_delim(Dates, delim = ", ") %>%
  select(Date = Dates, Customer, Product) %>%
  mutate(
    Date = as.character(Date),
    Date = case_when(
      str_detect(Date, "^\\d+$") ~ as.Date(
        as.numeric(Date),
        origin = "1899-12-30"
      ),
      TRUE ~ coalesce(dmy(Date), mdy(Date))
    )
  )

# transformation technically correct. one extra row in the result and incosistence in the date format.

File: files/300-399/390/CH-390.py
import pandas as pd

path = "300-399/390/CH-390 Table Transformation.xlsx"
input = pd.read_excel(path, usecols="B:E", nrows=4, skiprows=2)
test = pd.read_excel(path, usecols="G:I", nrows=7, skiprows=2)

input.columns = input.columns.astype(str)
input = input.rename(columns={input.columns[0]: "col1"})

long = input.melt(id_vars=["col1"], var_name="col", value_name="val")
wide = long.pivot_table(index="col", columns="col1", values="val", aggfunc="first").reset_index()
wide.columns.name = None
wide = wide.assign(Dates=wide["Dates"].astype(str).str.split(r",\s*"))

result = (
    wide.explode("Dates")
    .reset_index(drop=True)
    .loc[:, ["Dates", "Customer", "Product"]]
    .rename(columns={"Dates": "Date"})
)

s = result["Date"].astype(str).str.strip()
is_num = s.str.fullmatch(r"\d+")
date_num = pd.to_datetime(s.where(is_num).astype(float), unit="D", origin="1899-12-30", errors="coerce")
date_dmy = pd.to_datetime(s.where(~is_num), dayfirst=True, errors="coerce")
date_mdy = pd.to_datetime(s.where(~is_num), dayfirst=False, errors="coerce")
result["Date"] = date_num.fillna(date_dmy).fillna(date_mdy)

# transformation technically correct. one extra row in the result and incosistence in the date format.